pkg/kubectl/drain/examples.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package drain
+
+import (
+	"fmt"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// This file contains examples on how to drain/cordon/uncordon nodes.
+// These functions may be called directly, or their functionality
+// copied into your own code, for example if you want different output
+// behaviour.
+
+// RunNodeDrain shows the canonical way to drain a node.
+// You should first cordon the node, e.g. using RunCordonOrUncordon
+func RunNodeDrain(drainer *Helper, nodeName string) error {
+	list, errs := drainer.GetPodsForDeletion(nodeName)
```

Comment:
So, not 100% related to this patch set, but in relation to machine-api draining I'm seeing the following (IgnoreAllDaemonSets: true):

```
"ci-op-ftvj407t-8b330-w54nt-worker-us-east-1a-mpk6r": drain failed for node "[cannot delete daemonsets.apps \"machine-config-daemon\" is forbidden: User \"system:serviceaccount:openshift-machine-api:machine-api-controllers\" cannot get resource \"daemonsets\" in API group \"apps\" in the namespace \"openshift-machine-config-operator\": openshift-machine-config-operator/machine-config-daemon-ttwkq, cannot delete daemonsets.apps \"node-exporter\" is forbidden: User \"system:serviceaccount:openshift-machine-api:machine-api-controllers\" cannot get resource \"daemonsets\" in API group \"apps\" in the namespace \"openshift-monitoring\": openshift-monitoring/node-exporter-w7xsf, cannot delete daemonsets.apps \"multus\" is forbidden: User \"system:serviceaccount:openshift-machine-api:machine-api-controllers\" cannot get resource \"daemonsets\" in API group \"apps\" in the namespace \"openshift-multus\": openshift-multus/multus-84l5h, cannot delete daemonsets.apps \"ovs\" is forbidden: User \"system:serviceaccount:openshift-machine-api:machine-api-controllers\" cannot get resource \"daemonsets\" in API group \"apps\" in the namespace \"openshift-sdn\": openshift-sdn/ovs-h8vk2, cannot delete daemonsets.apps \"sdn\" is forbidden: User \"system:serviceaccount:openshift-machine-api:machine-api-controllers\" cannot get resource \"daemonsets\" in API group \"apps\" in the namespace \"openshift-sdn\": openshift-sdn/sdn-g2qnp, cannot delete daemonsets.apps \"tuned\" is forbidden: User \"system:serviceaccount:openshift-machine-api:machine-api-controllers\" cannot get resource \"daemonsets\" in API group \"apps\" in the namespace \"openshift-cluster-node-tuning-operator\": openshift-cluster-node-tuning-operator/tuned-d4gfr, cannot delete daemonsets.apps \"dns-default\" is forbidden: User \"system:serviceaccount:openshift-machine-api:machine-api-controllers\" cannot get resource \"daemonsets\" in API group \"apps\" in the namespace \"openshift-dns\": openshift-dns/dns-default-6xz6m, cannot delete daemonsets.apps \"node-ca\" is forbidden: User \"system:serviceaccount:openshift-machine-api:machine-api-controllers\" cannot get resource \"daemonsets\" in API group \"apps\" in the namespace \"openshift-image-registry\": openshift-image-registry/node-ca-6cc28]"; %!v(MISSING)
```

So, we should probably ignore daemonsets earlier in the process or ignore these errors.

Reply:
I'm not sure where that error comes from, but you're right that the permission thing is problematic.  I wonder if we should look at the ownerRef on pods to see if they are part of a daemonset instead.

Reply:
I have a patch out against this bug: https://github.com/kubernetes/kubernetes/pull/80129

So, I don't think you need to address it in this patch.  The problem was I running into was my service account had permissions for v1beta1 daemonsets but not v1.